src/librustc/infer/error_reporting/mod.rs

```diff
@@ -966,3 +712,515 @@ impl<'tcx> ObligationCause<'tcx> {
+        } else {
+            s
+        }
+    }
+
+    fn ty_str(&self, ty: Ty<'tcx>) -> String {
+        match ty.sty {
+            ty::TyAdt(def, sub) => {
+                let mut path = self.item_path_str(def.did.clone());
+
+                // Only draw `<...>` if there're lifetime/type arguments.
+                let len = sub.len();
+                if len > 0 {
+                    path.push_str("<");
+                }
+
+                let lifetimes = sub.regions()
+                    .map(|lifetime| self.lifetime_display(lifetime))
+                    .collect::<Vec<_>>()
+                    .join(", ");
+                if lifetimes.len() > 0 {
+                    path.push_str(&lifetimes);
+                }
+                let type_args = sub.types()
+                    .map(|ty| self.ty_str(ty))
+                    .collect::<Vec<_>>()
+                    .join(", ");
+                if lifetimes.len() > 0 && type_args.len() > 0 {
+                    path.push_str(", ");
+                }
+                if type_args.len() > 0 {
+                    path.push_str(&type_args);
+                }
+
+                // Close the type argument bracket.
+                // Only draw `<...>` if there're lifetime/type arguments.
+                if len > 0 {
+                    path.push_str(">");
+                }
+                path
+            }
+            ty::TyArray(ty, tyconst) => format!("[{}; {}]",
+                                                self.ty_str(ty),
+                                                tyconst.usize_val()),
+            ty::TySlice(ty) => format!("[{}]", self.ty_str(ty)),
+            ty::TyTuple(tys, _) => {
+                let tuple_tys = tys.iter().map(|ty| self.ty_str(ty)).collect::<Vec<_>>();
+                if tuple_tys.len() == 1 {
+                    format!("({},)", tuple_tys[0])
+                } else {
+                    format!("({})", tuple_tys.join(", "))
+                }
+            }
+            ty::TyRawPtr(ty_and_mut) => {
+                let mut s = String::new();
+                s.push_str("*");
+                if let hir::Mutability::MutMutable = ty_and_mut.mutbl {
+                    s.push_str("mut ")
+                } else {
+                    s.push_str("const ")
+                }
+                s.push_str(&self.ty_str(ty_and_mut.ty));
+                s
+            }
+            ty::TyRef(region, ty_and_mut) => {
+                let mut s = String::new();
+                s.push_str("&");
+                let r = region.to_string();
+                s.push_str(&r);
+                if !r.is_empty() {
+                    s.push_str(" ");
+                }
+                if let hir::Mutability::MutMutable = ty_and_mut.mutbl {
+                    s.push_str("mut ")
+                }
+                s.push_str(&self.ty_str(ty_and_mut.ty));
+                s
+            }
+            _ => format!("{}", ty),
```

Comment:
You really want to replace the entirety of ppaux here. Otherwise e.g. everything within a trait object will go back to ugly.

Reply:
Yes, that is indeed true. Given how big this PR was getting I thought that it was a good idea to put it out there to get feedback before going further down this rabbit hole. That being said, this covered a big enough percentage of cases that allowed me to test this approach.